### DISCUSSION TITLE
forge test -m  throws an  error!!!

### DISCUSSION THREAD

**Discussion by DCVglobalnetwork** [Cyfrin/foundry-full-course-cu#1648] [General]
each time i use forge test -m throws on error 

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/2634aafa-8e68-4bbd-96f8-ffd6abd7072f)

please anybody 
any idea why?

**Comment by EngrPips** [comment]
Hello @DCVglobalnetwork, please use `--mt` instead as `-m` is deprecated.

**Reply by DCVglobalnetwork** [reply]
Thank you for your reply 
now the function public is an issue . I tested with public and there was an error about new restrictions so I added view as view was mented. I sent the code and test . if you can help please. Thank you 

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/4cbd26b0-9d02-4a1b-9de8-45331cb1ac10)



![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/1dfd41b4-61c5-4d29-b4a1-ea321c0428bc)

**Reply by EngrPips** [reply]
Hello @DCVglobalnetwork, Please run `foundry` and also try prank `bob` as @Chinwuba22 has mentioned, and let's see how that will play out.

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/18d7d9f5-b0e6-4849-8c90-841ca5884cf5)
 

the same

**Reply by DCVglobalnetwork** [reply]
when you wrote run Foundry u mean install again ?

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/1366c56a-643a-4045-8bed-d822790eaaf9)

**Reply by EngrPips** [reply]
I meant `foundryUp`, Sorry for the confusion.

**Reply by DCVglobalnetwork** [reply]
Hi, yes I did run 
```shell
foundryup
``` 
and the same error

**Comment by drained69** [comment]
Prank any other address that is not `msg.sender` in your setUp. Prank 'bob` maybe.

**Comment by drained69** [comment]
Sorry @DCVglobalnetwork, would be nice to add `-vvvv` to your forge test, so you understand better the cause of the issue. 

I think the cause of this is the way you are going about the pranking still and not updating your foundry. For your .transfer to work the sender has to be the owner. In your own case above the deployer contract address is most likely the owner. You can copy out your code here also maybe that will help.

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/647c1feb-de7d-4482-8b94-796af6c2243e)


Today with forge test --mt worked for my nft peoject 
does not work for THIS ONE?

**Reply by DCVglobalnetwork** [reply]
```shell
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {DeployOurToken} from "../script/DeployOurToken.s.sol";
import {OurToken} from "../src/OurToken.sol";
import {Test} from "forge-std/Test.sol";

contract OurTokenTest is Test {
    OurToken public ourToken;
    DeployOurToken public deployer;

    address bob = makeAddr("bob");
    address alice = makeAddr("alice");

    uint256 public constant STARTING_BALANCE = 100 ether;

    function setUp() public {
        deployer = new DeployOurToken();
        ourToken = deployer.run();

        vm.prank(bob);
        ourToken.transfer(bob, STARTING_BALANCE);
    }

    function testBobBalance() public view {
        assertEq(STARTING_BALANCE, ourToken.balanceOf(bob));
    }
}
```

**Reply by EngrPips** [reply]
Hey @DCVglobalnetwork, The command you ran is `forge test -vvvv testBobBalance` instead of `forge test --mt testBobBalance -vvvv`

**Reply by DCVglobalnetwork** [reply]
yes I did run  

```shell
forge test --mt testBobBalance
```

it doesn't go through, I'm stuck in a loop
but for other project it works perfectly with 

```shell
forge test --mt
```

and the question is how? or why?

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/e245adcf-3c13-4535-a8fa-a12c47cfb2af)

**Reply by EngrPips** [reply]
Hello @DCVglobalnetwork, Please run the test with the `-vvvv` flag and show the test output. That way we would have more context and be able to debug it, but from the little context I have right now, something seems wrong in your test `setUp` as the error says `setUp failed.`

**Reply by DCVglobalnetwork** [reply]
I did try I started over again from the scratch and the same error 

i pushed the code 

https://github.com/DCVglobalnetwork/erc20-test.git

**Reply by DCVglobalnetwork** [reply]
and I run the test over again this time 

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/e000795d-08ed-42c6-b7f4-3b5179558b71)

**Reply by EngrPips** [reply]
@DCVglobalnetwork, Please run the test with the `-vvvv` flag and share the test output. That would help me debug this.

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/0af2b3e7-c044-4c59-9dfa-735a4a4e014c)

is that output you are asking ? 

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/27cacab3-fb67-412b-a469-791bfae11c40)

**Reply by DCVglobalnetwork** [reply]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/f5c806b7-6428-4886-b7b4-4275753387dd)


![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/d38f8540-6ae5-40fb-b43b-cab804547aa5)

**Reply by drained69** [reply]
You still havent pushed to github it would be easier for him to help you there.

**Comment by drained69** [comment]
push your code to github and share a link here.

**Reply by DCVglobalnetwork** [reply]
https://github.com/DCVglobalnetwork/erc20-test.git

**Comment by drained69** [comment]
Kindly recheck the file you sent. When pushing dont forget to use `git add .` before pushing the `.` is important.

**Reply by DCVglobalnetwork** [reply]
yes , I will do it as I will start over again and see where is the main error meanwhile I was trying to find what is happening and did another very simple project and they seems to be similar as function test for testBobBalance at the begining had the same issus 
Can u look at screenshot below , do you know how test can be passed ?


![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/74274fdc-8e50-42fc-b8ff-58db97fd68c6)

**Reply by EngrPips** [reply]
The test output stated that your assertion failed and it went further to explain that `HelloWorld != Hello World`

**Reply by EngrPips** [reply]
Make your `testGreet` function as below and rerun the test.

```js
function testGreet() public view {
assertEq(hellowWorld.greet(), "HelloWorld")
}
```

**Reply by DCVglobalnetwork** [reply]
thank you 
It did not work with hellowWorld as found it error so I used it 

```shell
function testGreet() public view {
assertEq(hellowWorld.greet(), "HelloWorld")
}
```
the error was with space between  "Hello World" you were right about "HelloWorld"

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/f2078ee6-4d59-4cde-8a63-505e6dee0eea)

**Reply by EngrPips** [reply]
Good you found the error, Keep crushing it.

**Reply by DCVglobalnetwork** [reply]
Hi 
I started over again this time I have new error 
here is the code 

```shell
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {OurToken} from "../src/OurToken.sol";

contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;

    function run() external {
        vm.startBroadcast();
        new OurToken(INITIAL_SUPPLY);
        vm.startBroadcast();
    }
}
```
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/2285441b-c68b-490c-8fb5-a0032ea2f2a4)

**Comment by DCVglobalnetwork** [comment]
I started over again 

```shell
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {OurToken} from "../src/OurToken.sol";

contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;

    function run() external {
        vm.startBroadcast();
        new OurToken(INITIAL_SUPPLY);
        vm.startBroadcast();
    }
}
```
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/2452058f-0ab2-4190-87a9-1b4c211a1b83)

**Reply by EngrPips** [reply]
Hello @DCVglobalnetwork, You started  broadcast with `vm.startBroadcast()` and you also ended it with `vm.startBroadcast()` instead of `vm.stopBroadcast()`

**Comment by DCVglobalnetwork** [comment]
passed the test

I'm doing it for the third time
Can anyone explain to me how this is possible? i did exactly the same way all of them 

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/105791829/2fecddf7-61bd-46ae-9a33-e8c5bd44da3d)

**Reply by EngrPips** [reply]
I am glad you finally crushed the test. You must be missing some details initially.

**Reply by DCVglobalnetwork** [reply]
I was able to complete the entrie project without any major issuse .
 Thank you for your 24/7 support

**Reply by EngrPips** [reply]
You are welcome. Keep crushing it.
